compiler/env/newmemory/DebugAllocator.cpp

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2017, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if (defined(LINUX) && !defined(OMRZTPF)) || defined(__APPLE__) || defined(_AIX)
+#include <sys/mman.h>
+#if defined(__APPLE__) || !defined(MAP_ANONYMOUS)
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+#elif defined(OMR_OS_WINDOWS)
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#else
+#include <string.h>
+#endif /* defined(OMR_OS_WINDOWS) */
+
+#include <stddef.h>  // for size_t
+#include <cstdlib>   // for free, malloc
+#include <new>       // for bad_alloc, nothrow, nothrow_t
+#include "infra/Assert.hpp"
+#include "env/mem/DebugAllocator.hpp"
+
+void *
+TR::DebugAllocator::allocate(size_t size, const std::nothrow_t tag,void * hint) const throw()
+   {
+#if (defined(LINUX) && !defined(OMRZTPF)) || defined(__APPLE__) || defined(_AIX)
+   void *p = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+   return (p == MAP_FAILED) ? NULL : p;
+#elif defined(OMR_OS_WINDOWS)
+   return  VirtualAlloc(NULL, size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+#else
+   return this->TR::RawAllocator::allocate(size, hint);
```

Comment:
I don't think this can work. If the intention is to cause a failure, can `static_assert` be used instead?

Reply:
That is another option, but the option still exists to paint the memory in `prrotect()` even if it won't fault on re-use. Asserting here would prevent that option.

Reply:
I'm not sure I follow. Calling `TR::RawAllocator::allocate(size_t, void *)` will call `TR::RawAllocator::allocate(size_t, nothrow_t, void*)`, which will be delegated back to `TR::DebugAllocator::allocate(size_t, nothrow_t, void *)`, resulting in an infinit loop... unless I've missed something?

There's also the issue of calling something that can throw an exception inside a `throw ()` function without handling the exception.

Reply:
yean, that's wrong. `TR::DebugAllocator `should extend `TR::MallocAllocator` and those calls should go back to `TR::MallocAllocator` <blush>

Reply:
good catch on the missing `tag` parameter...i've added it to the call. Thanks!